en/lessons/typeclasses/comparison.md

```diff
@@ -0,0 +1,590 @@
+
+Not everything can be tested for equality. For example, the `Eq` class does not include functions. So if we try to compare two functions, what we get is a compile error:
+
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> (\x -> x + 1 + 1) == (\x -> x + 2)
+
+error:
+    • No instance for (Eq (Integer -> Integer))
+```
+
+Although in this case we can see that these two functions are the same (they both add two to a number), this determination is not always possible to make. So, `(==)` is not defined for functions, and thus function types such as `Integer -> Integer` do not belong to the `Eq` class.
+
+[`IO`](https://hackage.haskell.org/package/base/docs/Prelude.html#t:IO) actions also do not have a meaningful equality test. Again we see compilation failure when we try to use the `(==)` operation with a type that does not support it:
+
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> putStrLn "hello" == writeFile "result.txt" "okay"
+error:
+    • No instance for (Eq (IO ())) arising from a use of ‘==’
+```
+
+### Deriving Eq for new types
+
+Suppose we define a `Message` as a product of two [`Text`](https://hackage.haskell.org/package/text/docs/Data-Text-Lazy.html#t:Text) values:
+
+```haskell
+import Data.Text.Lazy (Text)
+import qualified Data.Text.Lazy as Text
+
+data Message = Message { subject :: Text, body :: Text }
+
+message1 = Message{ subject = "Hi", body = "How are you"}
+message2 = Message{ subject = "Hi", body = "How are you???"}
+```
+
+The `Text` type belongs to the `Eq` class. This lets us test the two messages' subjects and bodies for equality.
+
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> subject message1 == subject message2
+True
+ghci> body message1 == body message2
+False
+```
+
+We might also want to ask whether the two messages are equal. However, we cannot yet, because the `(==)` function is not yet defined for our `Message` type:
+
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> message1 == message2
+error:
+    • No instance for (Eq Message) arising from a use of ‘==’
+```
+
+We do not need to define `(==)` explicitly, because the compiler can generate it for us. Modify the definition of the `Message` datatype by adding an additional clause that begins with the `deriving` keyword:
+
+```haskell
+data Message = Message { subject :: Text, body :: Text }
+    deriving Eq
+```
+
+This `Message` type now has an instance of `Eq`, and so its values can be tested for equality.
+
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> message1 == message2
+False
+ghci> message1 /= message2
+True
+```
+
+Obtaining an `Eq` instance in this way is called "stock deriving" because it takes advantage of the compiler's built-in support for deriving the `Eq` class.
+
+#### When Eq cannot be derived
+
+To derive a stock `Eq` instance for a product type like `Message`, all of the constructor's fields must have `Eq` instances. Suppose our type contains an additional field, an `IO` action which sends the message somewhere.
```

Comment:
Wouldn't mind changing it, but - I use "product" because it sounds related to "sum", and these two varieties of type are often discussed together in contrast. If we avoid saying "product" in favor of "record", is there also an alternative to "sum"?

Reply:
I think we can safely express the "intent" of a sum type by calling it a "variant".

Reply:
Other languages call them "tagged unions" as well, but I'm more in favour of "variant".